WireDomain/Sources/WireDomain/NotificationService/NotificationSession.swift

```diff
@@ -0,0 +1,97 @@
+//
+// Wire
+// Copyright (C) 2025 Wire Swiss GmbH
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Combine
+import WireAPI
+import WireDataModel
+
+/// Observes pending events, process them and generates new notifications content.
+final class NotificationSession {
+
+    // MARK: - Failure
+
+    enum Failure: Error {
+        case unableToPullPendingEvents(Error)
+    }
+
+    // MARK: - Properties
+
+    private let updateEventsRepository: any UpdateEventsRepositoryProtocol
+    private var subscription: AnyCancellable?
+
+    // MARK: - Object lifecycle
+
+    init(
+        updateEventsRepository: any UpdateEventsRepositoryProtocol,
+        onNotificationContent: @escaping (UNMutableNotificationContent) -> Void
+    ) {
+        self.updateEventsRepository = updateEventsRepository
+        self.subscription = updateEventsRepository.observePendingEvents()
+            .collect() // Collects all the events batches.
+            .map { $0.flatMap { $0 } }
+            .map(generateNotificationContent)
+            .sink(receiveValue: onNotificationContent)
+    }
+
+    deinit {
+        subscription?.cancel()
+        subscription = nil
+    }
+
+    // MARK: - Notifications
+
+    func processPushNotification(
+        eventID newEventID: UUID
+    ) async throws {
+        let lastEventId = updateEventsRepository.fetchLastEventEnvelopeID()
+
+        if lastEventId == nil {
+            updateEventsRepository.storeLastEventEnvelopeID(newEventID)
+        }
```

Comment:
If there's no last event id then it means we need to perform a slow sync, so I would throw an error here. In this case, we actually have no use for the `eventID` argument.